### ISSUE TITLE
Build next generation polyfill service using `core-js` polyfills

### ISSUE THREAD

**Issue by neenhouse** [zloirock/core-js#638]
# Overview / RFC

(Work in progress)

Looking forward, we'd like to build a next generation polyfill service that uses corejs polyfills as it's core to capture next gen performance.  As stated from @zloirock a look to the future:

> A service like this one integrated with a good polyfills source like core-js, which only loads the needed polyfills by statically analyzing the source like Babel's useBuiltIns: usage option does could cause a revolution in the way we think about polyfills.

We've tested and built this service and wish to open source it back to community with the help and input of @zloirock.  We should be able to share the details of the implementation in a RFC here and are actively working on open sourcing the internal components we have already built.  We intend to publish a demo of this service to `corejs.io` as a proof of concept and were hoping to co-maintain this package as an extension of corejs.io. 

Lots of details to document here and discussion to happen in the next few days.

**Comment by zloirock** [comment]
Sorry for the delayed reply, I have some problems with time. Sounds great. Waiting for updates...

**Comment by neenhouse** [comment]
Also apologize for the delay on our end.  We should have code published soon.

**Comment by slavafomin** [comment]
@neenhouse this looks extremely interesting, we would be very glad to see such service implemented. Do you have more progress? Is there a way for us to help?

**Comment by neenhouse** [comment]
@slavafomin Yes!  We fully intend to do all phases in the open.   Since releasing our code has taken longer than anticipated, I would be open to meeting to discuss our next steps.  I'm wondering if @zloirock would be open to meeting up with us as well?

**Comment by zloirock** [comment]
@neenhouse sorry for the delay, I had (and still have) some serious problems, but let's try to find a convenient time for both of us. Could you mail me a convenient time for you?

**Comment by Awen-hub** [comment]
> # Overview
> Looking forward, we'd like to build a next generation polyfill service that uses corejs polyfills as it's core to capture next gen performance. As stated from @zloirock a look to the future:
> 
> > A service like this one integrated with a good polyfills source like core-js, which only loads the needed polyfills by statically analyzing the source like Babel's useBuiltIns: usage option does could cause a revolution in the way we think about polyfills.
> 
> We've tested and built this service and wish to open source it back to community with the help and input of @zloirock. We should be able to share the details of the implementation in a RFC here and are actively working on open sourcing the internal components we have already built. We intend to publish a demo of this service to `corejs.io` as a proof of concept and we're hoping to co-maintain this package as an extension of corejs.io.
> 
> # Motivation
> World class performance starts with fully optimized critical load path. Polyfills are mandatory in the critical load path if you are building a modern UI application. A service oriented approach is the best way to optimize modern applications, and corejs is a world-class source of polyfills.
> 
> # Details
> In order to build a service that can distribute polyfill bundles, we need to build a service that can:
> 
> * Accept GET HTTP requests cross origin
> * Normalize user-agent header to small subset
> * Respond with an optimized javascript bundles based on core-js library
> 
> The challenge in implementing a service that can bundle and respond _quickly_. The approach to achieve this is inspired by polyfill.io. You mainly depend on caching responses in a CDN network to ultimately get the best performance that is close to the user. The second caching layer which this approach can improve on polyfill.io is using pre-computation in CDN compute services to optimize cold cache hits from the CDN.
> 
> The CDN cache layer should memoize input based on:
> 
> * normalized user agent string based on [browserslist](https://github.com/browserslist/browserslist) query
> * content-encoding (none, gzip, brotli)
> * query string params
>   
>   * features - The feature set of core-js poyfills to use
>   * version - The version of the resource to request
> 
> When a CDN does not have a cached resource, a pre-computed bundle should be made available based on a subset of the same memoization criteria above. We have successfully designed and deployed this code and open source is pending.
> 
> The DNS record we have registered is `corejs.io`. A prototype service will be built in open source as part of this proposal, but is pending a service library we have named `nimbuild` that enables the pre-computation of core-js polyfill bundles to make the service possible.
> 
> Apologize again for the long wait, our open source process is taking longer than anticipated to finish.
> 
> # Alternatives
> Polyfill.IO has greatly inspired this approach. We initially implemented the service, but found a few issues with performance and stability that ultimately led us to build a similar service with corejs polyfills as a source. We have opened PRs to polyfill-library to improve the quality of the lib and still are willing to support their effort, but we feel that starting with core-js polyfills as a source will most likely lead to the best solution in terms of both quality and performance.

@neenhouse  Two years later, I have the same idea as you mentioned. However, in my opinion, the quality of polyfill-library is also not good as core-js now and their team seems not having much time and manpower to update and maintain the service. Do you have any progress in this side now?

**Comment by dargmuesli** [comment]
@zloirock I've sent you an email on this topic :rocket: happy to hear your response so that I can into contact with others I mentioned in the mail :raised_hands:

@neenhouse could you shed some light on how your plan developed? :pray:
